### PR TITLE
soc: intel_adsp: cavs: add simple imr functionality

### PR DESCRIPTION
**Pull request by juimonen**
Add simple imr mechanism to load the image from rom instead of doing cold boot. Basically we are only setting a flag for the boot to load existing image from imr area.

### CODE REVIEW

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/* Copyright (c) 2023 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_SOC_INTEL_ADSP_TGL_IMR_LAYOUT_H_
+#define ZEPHYR_SOC_INTEL_ADSP_TGL_IMR_LAYOUT_H_
+
+/* These structs and macros are from from the ROM code header
```
Comment (kv2019i) [inline soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h:7]: "from from"

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/* Copyright (c) 2023 Intel Corporation
```
Comment (kv2019i) [inline soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h:1]: - IMR with capitals (Isolated Memory Region)
- "to load image from IMR memory"

**Review by kv2019i — CHANGES_REQUESTED** [review]
Change looks good, a few minor comments on comments.

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h (hunk 1 above)
Comment (juimonen) [reply soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h]: ok

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h (hunk 2 above)
Comment (juimonen) [reply soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h:1]: ok tried to make the commit message better...

**Review by jxstelter — DISMISSED** [review]

**Review by tmleman — DISMISSED** [review]

**Review by softwarecki — DISMISSED** [review]

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+/* These structs and macros are from the ROM code header
+ * on cAVS platforms, please keep them immutable
+ */
+
+/*
+ * A magic that tells ROM to jump to imr_restore_vector instead of normal boot
+ */
+#define ADSP_IMR_MAGIC_VALUE		0x02468ACE
+#define IMR_L1_CACHE_ADDRESS		0xB0000000
```
Comment (andyross) [inline soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h:15]: Nitpick: memory layout constants belong in devicetree, not C code.  Pretty sure there's already a block defined there, no?

**Review by andyross — COMMENTED** [review]
Is this just a noop?  It enables the soft/IMR boot functionality but just jumps to the normal OS startup?  Why bother?  Shouldn't there be an API somewhere like "intel_adsp_set_imr_restore_vector()" or whatever?

**Review by andyross — COMMENTED** [review]
Few more nitpicks

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h (hunk 3 above)
Comment (kv2019i) [reply soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h:15]: @andyross FYI, this cavs25 version is more-or-less copy exact from the Intel ace version  (soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/adsp_imr_layout.h).

**Comment by kv2019i** [comment]
@andyross wrote:
> Is this just a noop? It enables the soft/IMR boot functionality but just jumps to the normal OS startup? 

No, this does enable significantly faster boot as the authenticated FW image already in IMR can be used for booth (imr_restore_vector is set). This does not restore state, but allows for much faster boot.

**Review by kv2019i — DISMISSED** [review]
Thanks @juimonen , looks good now.

I'd prefer to copy the ace15 implementation (that's in the tree already) as close as possible.

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h (hunk 3 above)
Comment (juimonen) [reply soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h]: now put to device tree

File: soc/xtensa/intel_adsp/cavs/power.c [hunk 4]
```diff
@@ -81,6 +87,13 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		soc_cpus_active[cpu] = false;
 		sys_cache_data_flush_and_invd_all();
 		if (cpu == 0) {
+			struct imr_layout *imr_layout = (struct imr_layout *)(IMR_L1_CACHE_ADDRESS);
+
+			imr_layout->imr_state.header.adsp_imr_magic = ADSP_IMR_MAGIC_VALUE;
+			imr_layout->imr_state.header.imr_restore_vector =
+					(void *)rom_entry;
+			sys_cache_data_flush_range(imr_layout, sizeof(*imr_layout));
```
[the remark this answers is not in the document]
Comment (juimonen) [reply soc/xtensa/intel_adsp/cavs/power.c]: ok

File: soc/xtensa/intel_adsp/cavs/power.c [hunk 5]
```diff
@@ -81,6 +87,13 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		soc_cpus_active[cpu] = false;
 		sys_cache_data_flush_and_invd_all();
 		if (cpu == 0) {
+			struct imr_layout *imr_layout = (struct imr_layout *)(IMR_L1_CACHE_ADDRESS);
+
+			imr_layout->imr_state.header.adsp_imr_magic = ADSP_IMR_MAGIC_VALUE;
+			imr_layout->imr_state.header.imr_restore_vector =
+					(void *)rom_entry;
```
[the remark this answers is not in the document]
Comment (juimonen) [reply soc/xtensa/intel_adsp/cavs/power.c]: ok

File: soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+struct imr_header {
+	uint32_t adsp_imr_magic;
+	uint32_t structure_version;
+	uint32_t structure_size;
+	uint32_t imr_state;
+	uint32_t imr_size;
+	void *imr_restore_vector;
```
[the remark this answers is not in the document]
Comment (juimonen) [reply soc/xtensa/intel_adsp/cavs/include/intel_tgl_adsp/adsp_imr_layout.h]: function pointer done. Unfortunately I have no idea why this struct is how it is and where it is originally documented. Just moving it from sof side...

**Comment by juimonen** [comment]
@andyross tried to address most of your comments... I'm trying to move cavs (tgl) in sof side to use the same platform file/files as ace variants, so that we would use as much as possible the same code and variation would be in dt. Just noticed then that I'm missing the IMR thing from power.c. 

If you look at the ace/power.c variant, you see it has like 2 implementations of IMR behind ifdefs: one "simple" and the other "with steroids". In cavs sof we only used the "simple" version, which only loads the authenticated image from IMR, saving some cycles between the linux driver and sof. When using sof in linux, for example in simple streaming use case, you don't too much notice this other than dmesg telling you sof is loading from IMR or not. The benefits are perhaps more noticeable when stressing the system heavily (starting/stopping multiple streams etc).

I could not find any documentation of the unused layout struct members, so I'm basically just blindly copying those over from sof side.

**Comment by marc-hb** [comment]
> I could not find any documentation of the unused layout struct members, so I'm basically just blindly copying those over from sof side.

Then please add in the comments a clear git reference pointing at where they came from.

**Review by andyross — DISMISSED** [review]

**Review by kv2019i — DISMISSED** [review]

**Review by ceolin — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by ceolin — DISMISSED** [review]

**Review by softwarecki — DISMISSED** [review]

**Comment by juimonen** [comment]
rebased.

**Review by kv2019i — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**